File: 프로그래머스/2/12953. N개의 최소공배수/N개의 최소공배수.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int solution(vector<int> arr) { // n개의 수의 최소공배수 구하기
    int answer = 1;
    
    vector<int> arr2;
    long long multi = 1; // 나누는 값
    for(int i = 2; i <= *max_element(arr.begin(), arr.end()); i++){ // 10
        
        while(true){
            int avail = 0;  arr2 = arr;  // 초기화
            for(int a = 0; a < arr.size(); a++){ // 15
                if(arr[a] % i == 0){
                    arr2[a] = arr[a] / i; // 몫 저장
                    avail++;
                }
            }
            if(avail > 0){
                arr = arr2; multi *= i;
            } else break;
        } 
    }
    
    for(int i = 0; i < arr.size(); i++){
        answer *= arr[i];
    }
    answer *= multi; 
    
    return answer;
}

File: 프로그래머스/2/12953. N개의 최소공배수/N개의 최소공배수.java
import java.util.*; 

class Solution {
    public int solution(int[] arr) {
        int answer = 1;
        boolean[] finish = new boolean[arr.length]; 
        int fin = 0; 
        // 하나씩 1로 만들기
        for(int n = 2; n < 100; n++){
            boolean avail = false; 
            for(int i = 0; i < arr.length; i++){
                if(finish[i]) continue; 
                if(arr[i] % n == 0){
                    avail = true; 
                    arr[i] /= n; 
                    if(arr[i] == 1) {
                        finish[i] = true; fin++;
                    }
                }
            }
            if(avail) {
                answer *= n; 
                n--; // 다시
            }
            if(fin == arr.length) break; 
        }
        
        // n개의 숫자를 담은 배열 arr이 입력되었을 때 이 수들의 최소공배수
        return answer;
    }
}